File: solutions/elixir/roman-numerals/2/lib/roman_numerals.ex
defmodule RomanNumerals do
  @spec format(integer, integer, String.t(), String.t(), String.t()) :: String.t()
  defp format(number, base, ones, fives, tens) do
    number
    |> rem(10 * base)
    |> div(base)
    |> (fn n ->   
      cond do
        n == 0 -> ""
        n < 4 -> String.duplicate(ones, n)
        n == 4 -> ones <> fives
        n == 9 -> ones <> tens
        true -> fives <> String.duplicate(ones, n - 5)
      end
    end).()
  end


  @doc """
  Convert the number to a roman number.
  """
  @spec numeral(pos_integer) :: String.t()
  def numeral(number) do
      format(number, 1000, "M", " ", " ") <>
       format(number, 100, "C", "D", "M") <>
       format(number, 10, "X", "L", "C") <>
       format(number, 1, "I", "V", "X")
  end
end

File: solutions/julia/roman-numerals/1/roman-numerals.jl
function format(number::Int, unit::Int, ones::Char, fives::Char, tens::Char)
    number = div(number % (unit * 10), unit)
    if number < 4
        return ones ^ number
    elseif number == 4
        return ones * fives
    elseif number == 9
        return ones * tens
    elseif number >= 5
        return fives * (ones ^ (number - 5))
    else
        return ""
    end
end


function to_roman(number)
    if number <= 0
        throw(ErrorException("Error"))
    end
    out = ""
    out *= format(number, 1000, 'M', ' ', ' ')
    out *= format(number, 100, 'C', 'D', 'M')
    out *= format(number, 10, 'X', 'L', 'C')
    out *= format(number, 1, 'I', 'V', 'X')

    return out
end
